Sources/NIOHTTPCompression/HTTPResponseDecompressor.swift

```diff
@@ -130,40 +124,55 @@ public final class NIOHTTPResponseDecompressor: ChannelDuplexHandler, RemovableC
 
             context.fireChannelRead(data)
         case .body(var part):
-            switch self.state {
+            switch self.decompressor.state {
             case .compressed(_, let originalLength):
                 while part.readableBytes > 0 {
+                    var buffer = context.channel.allocator.buffer(capacity: 16384)
                     do {
-                        var buffer = context.channel.allocator.buffer(capacity: 16384)
-                        try self.stream.inflatePart(input: &part, output: &buffer)
-                        self.inflated += buffer.readableBytes
-
-                        if self.limit.exceeded(compressed: originalLength, decompressed: self.inflated) {
-                            context.fireErrorCaught(DecompressionError.limit)
-                            return
-                        }
-
-                        context.fireChannelRead(self.wrapInboundOut(.body(buffer)))
+                        try self.decompressor.decompress(part: &part, buffer: &buffer, originalLength: originalLength)
                     } catch {
                         context.fireErrorCaught(error)
                         return
                     }
+
+                    context.fireChannelRead(self.wrapInboundOut(.body(buffer)))
                 }
             default:
                 context.fireChannelRead(data)
             }
         case .end:
-            switch self.state {
+            switch self.decompressor.state {
             case .compressed:
-                inflateEnd(&self.stream)
+                self.decompressor.deinitializeDecoder()
             default:
                 break
             }
             context.fireChannelRead(data)
         }
     }
+}
+
+fileprivate struct Decompressor {
```

Comment:
could we also move this to the shared file?

Reply:
done!